Hearing post put

stephawe opened the pull request.

Commit:
Add a simple fixture for a Label

Commit:
Refactor and update the HearingCreateUpdateSerializer

Add a function _create_or_update_sections
Use this function in both create and update

Commit:
Fix import in test_comment.py

Commit:
Improve the display of errors in assert_common_keys_equal

Commit:
Fix the User.get_default_organization

Commit:
Add John Smith, a test user from an organization

Commit:
Update the hearing serializer to check the permissions

Only users from the same organization can update a hearing
Remove duplicated import Response
Expand Organization upon deserializing created/updated hearing
Return 'created_at' when deserializing created/updated hearing
Assign Organization while POSTing a hearing

Commit:
Move the test_image to tests/utils.py

Move the test_image into tests/utils.py
Add a function image_test_json which returns this image
Update the test_comment to use this function

Commit:
Add some tests for POSTing and PUTing hearings

tuomas777 commented on democracy/tests/conftest.py:
```diff
@@ -80,6 +80,12 @@ def default_hearing(john_doe, contact_person):
 
 @pytest.fixture()
+def default_label():
+    label = Label.objects.create(id=1, label='The Label')
```
Setting an id explicitly here probably isn't a good idea as it might lead to collisions afterwards

stephawe replied:
It is indeed a terrible idea!

tuomas777 commented on democracy/tests/test_hearing.py:
```diff
@@ -450,3 +535,227 @@ def test_hearing_filters(admin_api_client, default_hearing, updates, filters, ex
+    assert_hearing_equals(data, valid_hearing_json, john_smith_api_client.user)
+
+
+# Test that a user can update a hearing
```
This comment doesn't really seem to match the test? :)

stephawe replied:
oops

tuomas777 commented on democracy/tests/test_hearing.py:
```diff
@@ -450,3 +535,227 @@ def test_hearing_filters(admin_api_client, default_hearing, updates, filters, ex
+    assert data == {'status': 'User without organization cannot PUT hearings.'}
+
+
+# Test that a user can update a hearing
```
The same as above

tuomas777 commented on democracy/tests/test_hearing.py:
```diff
@@ -450,3 +535,227 @@ def test_hearing_filters(admin_api_client, default_hearing, updates, filters, ex
+        "servicemap_url": "url",
+    })
+    data['sections'][0]['title'] = 'First section'
+    data['sections'][1]['images'][0]['caption'] = 'New image caption'
```
Is this change actually verified somewhere?

stephawe replied:
it is now :)

tuomas777 commented on democracy/tests/test_hearing.py:
```diff
@@ -450,3 +535,227 @@ def test_hearing_filters(admin_api_client, default_hearing, updates, filters, ex
+    hearing = create_hearings(1)[0]
+    response = john_smith_api_client.post(endpoint, data=valid_hearing_json, format='json')
+    data = get_data_from_response(response, status_code=201)
+    data['sections'][0]['id'] = hearing.id
```
Would be a bit better, if the id was actually from an existing section id from another hearing

tuomas777 commented on democracy/views/hearing.py:
```diff
@@ -332,4 +358,16 @@ def map(self, request):
-        return Response(serializer.data)
+        return response.Response(serializer.data)
+
+    def create(self, request, pk=None):
```
I think `create()` will never get param `pk`

tuomas777 reviewed (commented):
Just some minor comments, overall very nice work!!

Commit:
Allow user to GET unpublished hearings from their own organization

Allow to GET unpublished hearings if same organization
Add tests

Commit:
Remove the unused section field "created_by" from section serializer

Commit:
Add a method is_visible_for in Hearing

Commit:
Test that unpublished section can be fetched

Commit:
Update the filter_by_hearing_visible

Unpublished Hearing are now visible to users of the same organization
Function can now be used for filtering Hearings with hearing_lookup=''

Commit:
Add PATCH Hearing support

Allow to update partially a Hearing using PATCH
Add test

Commit:
Add force_closed to the hearing serializer

Commit:
Remove unused import
